### PR TITLE
Add Raven, and refactor a bit

### PR DESCRIPTION
**Pull request by carolynjoneslee**
#### What’s this PR do?
This PR adds Raven.js, so that we can log errors to Sentry. It also includes some refactoring, to make the code a bit more robust and easier to work with.

#### Any background context you want to provide?
We've had trouble helping customers debug issues in the past because we had no error logs. This will help with debugging, and it will also help us know how often people are encountering errors.

There's still more refactoring to do, particularly in the Superwidget script (yesgraph-invites.js) so that we can more easily run unit tests. Those changes are excluded only so this PR doesn't get too huge.

#### Where should the reviewer start?
- The old `configureAPI()` function was replaced with the new `YesGraphAPIConstructor()`, so take a look at that.
- Also, there's a new `loadRaven()` function that (obviously) loads Raven.

#### Does this require changes on the API side?
Nope

#### How should this be manually tested?
To run our unit tests: `node_modules/karma/bin/karma start karma.conf.js --single-run`

#### What are the relevant tickets?
https://app.asana.com/0/77757700101210/136183303181172

#### Screenshots (if appropriate)
N/A

#### Does the knowledge base need an update?
Nope

### CODE REVIEW

File: dev/yesgraph.js [hunk 1]
```diff
@@ -49,256 +207,132 @@
...
-                    },
-                    "timestamp": new Date(),
-                    "type": "screen"
-                }, ]
-            };
-            return hitAPI(ANALYTICS_ENDPOINT, "POST", eventData);
-        }
+        this.noConflict = function() {
+            delete window.YesGraphAPI;
+            return self
```
Comment (kendallc) [inline dev/yesgraph.js]: missing a semicolon

File: dev/yesgraph.js [hunk 2]
```diff
@@ -49,256 +207,132 @@
...
+            readCookie: function(key) {
+                var key = key + "=";
+                var cookies = document.cookie.split(';');
+                for (var i = 0; i < cookies.length; i++) {
+                    var cookie = cookies[i];
+                    while (cookie.charAt(0) == ' ') cookie = cookie.substring(1);
+                    if (cookie.indexOf(key) == 0) {
+                        var value = cookie.substring(key.length, cookie.length);
+                        if (value != "undefined") return value;
+                    };
```
Comment (kendallc) [inline dev/yesgraph.js]: Does a function return "null" if there is no return? Might want to return something in the default case.

File: dev/yesgraph.js [hunk 3]
```diff
@@ -49,256 +207,132 @@
...
+                self.clientToken = data.token;
+                self.utils.setCookie('yg-client-token', data.token);
+            },
+            getOrFetchClientToken: function (userData) {
+                var data = {
+                    appName: YesGraphAPI.app
+                };
+                data.userData = userData || undefined;
+                self.clientToken = self.utils.readCookie('yg-client-token');
+                data.token = self.clientToken || undefined;
```
Comment (kendallc) [inline dev/yesgraph.js]: See comment above -- if we return something in the "readCookie" funciton then we don't need to do the "|| undefined" here

File: dev/yesgraph.js [hunk 4]
```diff
@@ -49,256 +207,132 @@
...
-                    if (target.length > 0) {
-                        initDeferred.resolve(target);
-                    }
-                }, 100);
-            initDeferred.always(function(){ clearInterval(timer); });
-            return initDeferred.promise();
-        }
+        this.utils = {
+            readCookie: function(key) {
+                var key = key + "=";
```
Comment (kendallc) [inline dev/yesgraph.js]: Does this modify the input parameter? Might want to rename the local variable to something else (Vim suggested this)

File: dev/yesgraph.js (hunk 3 above)
Comment (carolynjoneslee) [inline dev/yesgraph.js]: The `readCookie` function actually returns `undefined` if no match was found, so this is redundant. I can return that explicitly above for clarity, and then take it out here.

File: dev/yesgraph.js (hunk 4 above)
Comment (carolynjoneslee) [inline dev/yesgraph.js]: Yeah this isn't good, I'll change it

**Comment by kendallc** [comment]
Missing a semicolon here: https://github.com/YesGraph/yesgraph-superwidget/blob/f9652801235e94086a5d867f570ed4c01eb4a995/dev/yesgraph-invites.js#L223

**Comment by kendallc** [comment]
And line https://github.com/YesGraph/yesgraph-superwidget/blob/f9652801235e94086a5d867f570ed4c01eb4a995/dev/yesgraph-invites.js#L315

**Comment by kendallc** [comment]
There are a couple more places missing semicolons, or with an unnecessary semicolon (after if-else statements) according to Vim. Might want to run the code through a linter just to find them all.

**Comment by kendallc** [comment]
Other than the syntax errors, looks good. Coverage is up. :)

**Pull request merged**